pkg/providers/zone/zone.go

```diff
@@ -0,0 +1,198 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package zone
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// We don't want to retry too aggressively here because this API is somewhat slow,
+// but at the same time we want to wake back up eventually and try again in the case of an outage.
+// These values were picked somewhat arbitrarily to achieve that.
+const (
+	maxFailuresPerWindow = 10
+	windowBackoff        = 60 * time.Minute
+)
+
+type Clock interface {
+	Now() time.Time
+}
+
+// SubscriptionsAPI defines the interface for Azure Subscriptions client operations
+type SubscriptionsAPI interface {
+	NewListLocationsPager(
+		subscriptionID string,
+		options *armsubscriptions.ClientListLocationsOptions,
+	) *runtime.Pager[armsubscriptions.ClientListLocationsResponse]
+}
+
+// Provider handles zone support detection for Azure regions
+type Provider struct {
+	subscriptionsAPI SubscriptionsAPI
+	subscriptionID   string
+	clock            Clock
+
+	// Cached zone support data - maps region name to zone support boolean
+	zoneSupport map[string]bool
```

Comment:
Any reason to keep the whole mapping, rather than just the region we are running in? AKS clusters are inherently single region, right?